lexer/lex.go

```diff
@@ -5,9 +5,164 @@ import (
 	"github.com/NeowayLabs/abad/token"
 )
 
-type (
-	Tokval struct {
-		Type  token.Type
-		Value utf16.Str
+type Tokval struct {
+	Type  token.Type
+	Value utf16.Str
+}
+
+func (t Tokval) Equal(other Tokval) bool {
+	return t.Type == other.Type && t.Value.Equal(other.Value)
+}
+
+// Lex will lex the given crappy JS code (utf16 yay) and provide a
+// stream of tokens as a result (the returned channel).
+//
+// The caller should iterate on the given channel until it is
+// closed indicating a EOF (or an error). Errors should be
+// handled by checking the type of the token.
+//
+// A goroutine will be started to lex the given code, if you
+// do not iterate the returned channel the goroutine will leak,
+// you MUST drain the provided channel.
+func Lex(code utf16.Str) <-chan Tokval {
+	tokens := make(chan Tokval)
+	
+	go func() {
+	
+		currentState := lexerInitialState(code)
+		
+		for currentState != nil {
+			token, newState := currentState()
+			if token != nil {
```

Comment:
I failed to understand why states must return a pointer to Tokval. 
As far as I understand, every state *must* return a token (it cannot be nil).

Reply:
Yeah this is the part of the current design  that I got uneasy. If I do the way you say right now the returned state is always the initialState...but I think that with more complex scenario this will change...but then it seemed like something was wrong, so I allowed state changes that don't produce tokens. On the design that Rob Pike proposes it is harder to see if it works like that because each state chances returns just  a state and the token generation is all internal to the state functions, so in his design the state change could produce no token, one token or a hundred tokens x_x. But I think always returning a token seems better in the end, was just confused while I navigated the solution space and stuff